src/specific/s_audio_stream.c

```diff
@@ -386,21 +393,16 @@ bool S_Audio_StreamSoundClose(int sound_id)
 
     if (stream->sdl.stream) {
         SDL_FreeAudioStream(stream->sdl.stream);
-        stream->sdl.stream = NULL;
     }
 
-    stream->is_read_done = true;
-    stream->is_used = false;
-    stream->is_playing = false;
-    stream->is_looped = false;
-    stream->volume = 0.0f;
-    stream->timestamp = 0;
-    SDL_UnlockAudioDevice(g_AudioDeviceID);
-
     if (stream->finish_callback) {
         stream->finish_callback(sound_id, stream->finish_callback_user_data);
     }
 
+    S_Audio_StreamSoundClear(stream);
```

Comment:
Haven't tested it, but IMO this will now freeze the game when the BGM is about to be replayed, and crash it after a short while with a stack call overflow error:

1. S_Audio_StreamSoundClose
2. stream->finish_callback
3. Music_StreamFinished
4. Music_PlayLooped
5. Music_StopActiveStream
6. S_Audio_StreamClose
7. stream->finish_callback (which still isn't NULLed)
8. repeat ad infinitum

Even if it works due to some internal checks inside music.c, I think it's by accident, and there's still the issue of nesting SDL_LockAudioDevice calls. We can do better.

I suggest attaching the callback to a variable like so, and then calling it afterwards:

```c
void (*finish_callback)()&(*@++__ilovec) = stream->finish_callback;
void *finish_callback_user_data = stream->finish_callback_user_data;
S_Audio_StreamSoundClear(stream);
SDL_UnlockAudioDevice(g_AudioDeviceID);
if (finish_callback) {
    finish_callback(sound_id, finish_callback_user_data);
}
```

and sprinkle 2 or 3 blank lines for extra prettiness. (This is pseudocode in case anyone wonders.)

Reply:
This seems for convoluted to me than the original code. 😰

Reply:
@lahm86 WDYT

Reply:
The pseudocode above looks good and I think it makes sense to avoid the issue you mention (although I have tested it out and you're right, it doesn't appear to be a problem - but at least this should future-proof it).

Reply:
Thanks for the help Lahm. Seems to work! Last question is, `finish_callback_user_data` is never set to `NULL` like `finish_callback`. Should it be? (For the record, this was not the case in the code before this PR either. I just wanted to ask while we are touching this code.)

Reply:
Yeah let's clear that, too. (Inside `S_Audio_StreamSoundClear` of course.)